### ISSUE TITLE
Managing relationships and N + 1

### ISSUE THREAD

**Issue by zachdaniel** [ash-project/ash#303]
Right now, managing relationships effectively have an N+1 problem, if you nest them. I'm not sure what these optimizations would ultimately look like, but I think that we can do a fair amount of optimization here.

**Comment by dolfinus** [comment]
Hi.

Most of GraphQL frameworks introduce a `dataloader` concept which should handle this issue by taking ids of records that need to be fetched, and then fetching them in one batch.

Absinthe have https://github.com/absinthe-graphql/dataloader library for that. Is it possible to reuse it?

**Comment by zachdaniel** [comment]
Yeah, we use dataloader in `AshGraphql`, but this is a bit of a different problem space. This is for making "changes" to relationships, e.g submitting a change like:

```elixir
%{
  posts: [
    %{comments: [%{likes: [....]}, ....], 
    ....
  ]
}
```

**Comment by zachdaniel** [comment]
Ash read actions also behave in "roughly" the same way that data loader behaves, although two requests to the same resource (e.g if something is related via two different relationships), will be run separately, it does not run one request per related entity.

**Comment by zachdaniel** [comment]
We will solve for N+1 here after we introduce bulk actions, and managed relationships will be implemented via bulk updates/creates/destroys.

**Comment by zachdaniel** [comment]
I will add this as an item on the bulk actions issue.
